### DISCUSSION TITLE
On demand loading tables

### DISCUSSION THREAD

**Discussion by WyattLiu** [duckdb/duckdb#9159] [Q&A]
May I ask how to write SQL query so the underlying big file (csv/parquet) can be on-demand loaded?

I am testing out duckdb with TPCH Queries. Particularly Q5:

`<Schema> CREATE TABLE statements`
```
COPY CUSTOMER FROM './customer.tbl' (DELIMITER '|');
COPY LINEITEM FROM './lineitem.tbl' (DELIMITER '|');
COPY NATION FROM './nation.tbl' (DELIMITER '|');
COPY ORDERS FROM './orders.tbl' (DELIMITER '|');
COPY PART FROM './part.tbl' (DELIMITER '|');
COPY PARTSUPP FROM './partsupp.tbl' (DELIMITER '|');
COPY REGION FROM './region.tbl' (DELIMITER '|');
COPY SUPPLIER FROM './supplier.tbl' (DELIMITER '|');
```
`<Q5>`

Using the conventional COPY approach, even the table is not used, I have to load them and load them fully which has memory and run time implications, may I know what is the best approach to just "point" the table and run a query from there so the table is on-demand loaded. The stage of loading table should be in seconds if the "pointer" to tables works. 

Thank you for your help!

**Comment by Tishj** [comment]
You could make a view over the parquet file instead, then the file doesnt get scanned unless you query it

**Reply by WyattLiu** [reply]
Thank you for your input, could you please give an example? My naive thought is like below:
Right now I converted the tbl/csv into a parquet file using pandas but since there no column names from TPCH tbl files, I am honestly need a way that tells duckdb a parquet file is meant for a table and please only load when you need and load partially if possible.

Thank you again for your time

**Reply by Tishj** [reply]
https://duckdb.org/docs/sql/statements/create_view

**Reply by WyattLiu** [reply]
I tried with views, it looks like I no longer need create table statements. However, using plain CREATE VIEW customer_view AS SELECT * FROM read_parquet('/home/wyatt/10G_q5/customer.parquet'); seems not providing too much benefit since I directly use the view right away in the query. May I ask if read_parquet('/home/wyatt/10G_q5/customer.parquet') would enable the on-demand benefit from the parquet file?

**Reply by Tishj** [reply]
This is confusing
Error: Referenced column "" not found in FROM clause!

That doesn't look like the actual error, I don't think we allow empty column names?

But it looks like you'll need to specify the schema in your view query

Create view view_name as select <original_name>::<cast_type> as <new_name>, etc.. from ...

**Reply by WyattLiu** [reply]
I realized what I did wrong, as the default tpch tool generated tbl does not have column names, I used pa and pd to push down the column names and types. However, I wonder if this is sufficient to take advantage that we only read what we need inside the view?
```
Sorry for the edit: the previous reply now becomes:
I tried with views, it looks like I no longer need to create table statements. 
However, using plain 

CREATE VIEW customer_view AS SELECT * FROM read_parquet('/home/wyatt/10G_q5/customer.parquet'); 

does not provide too much benefit since I directly use the view right away in the query. 
May I ask if read_parquet('/home/wyatt/10G_q5/customer.parquet') would enable the on-demand benefit from the parquet file (partial read)?
```

**Reply by Tishj** [reply]
I think we perform both projection and filter pushdown for parquet files, so yes we only read the columns that are required

**Reply by WyattLiu** [reply]
Okay, I will try something more significant and take some measurements. Thank you.
The TPCH tool only generates tbl without column names. To cook a sound parquet file needs some hands-on work.
